pkg/sync/parallel_syncer.go

```diff
@@ -0,0 +1,1706 @@
+	HasExternalResources bool   // Whether external resources were found
+	ShouldFetchRelated   bool   // Whether related resources should be fetched
+}
+
+// LocalStateContext provides local state management for parallel syncer
+// This implements StateInterface without sharing the global state machine
+type LocalStateContext struct {
+	resourceID           *v2.ResourceId
+	pageToken            string
+	needsExpansion       bool
+	hasExternalResources bool
+	shouldFetchRelated   bool
+}
+
+// NewLocalStateContext creates a new local state context for a resource
+func NewLocalStateContext(resourceID *v2.ResourceId) *LocalStateContext {
+	return &LocalStateContext{
+		resourceID:           resourceID,
+		pageToken:            "",
+		needsExpansion:       false,
+		hasExternalResources: false,
+		shouldFetchRelated:   true, // Default to true for backward compatibility
+	}
+}
+
+// PageToken returns the current page token for this resource
+func (lsc *LocalStateContext) PageToken(ctx context.Context) string {
+	return lsc.pageToken
+}
+
+// NextPage updates the page token for the next page
+func (lsc *LocalStateContext) NextPage(ctx context.Context, pageToken string) error {
+	lsc.pageToken = pageToken
+	return nil
+}
+
+// SetNeedsExpansion marks that grant expansion is needed
+func (lsc *LocalStateContext) SetNeedsExpansion() {
+	lsc.needsExpansion = true
+}
+
+// SetHasExternalResourcesGrants marks that external resources were found
+func (lsc *LocalStateContext) SetHasExternalResourcesGrants() {
+	lsc.hasExternalResources = true
+}
+
+// ShouldFetchRelatedResources returns whether related resources should be fetched
+func (lsc *LocalStateContext) ShouldFetchRelatedResources() bool {
+	return lsc.shouldFetchRelated
+}
+
+// ParallelSyncConfig holds configuration for parallel sync operations
+type ParallelSyncConfig struct {
+	// Number of worker goroutines to use for parallel processing
+	WorkerCount int
+	// Default bucket for resource types not explicitly configured
+	// If a resource type doesn't specify a sync_bucket, this default will be used
+	DefaultBucket string
+}
+
+// DefaultParallelSyncConfig returns a default configuration
+func DefaultParallelSyncConfig() *ParallelSyncConfig {
+	return &ParallelSyncConfig{
+		WorkerCount:   1,
+		DefaultBucket: "", // Empty string means each resource type gets its own bucket
+	}
+}
+
+// WithWorkerCount sets the number of worker goroutines
+func (c *ParallelSyncConfig) WithWorkerCount(count int) *ParallelSyncConfig {
+	if count > 0 {
+		c.WorkerCount = count
+	}
+	return c
+}
+
+// WithDefaultBucket sets the default bucket for resource types that don't specify a sync_bucket
+func (c *ParallelSyncConfig) WithDefaultBucket(bucket string) *ParallelSyncConfig {
+	c.DefaultBucket = bucket
+	return c
+}
+
+// task represents a unit of work for the parallel syncer
+type task struct {
+	Action       Action
+	ResourceID   string
+	Priority     int              // Higher priority tasks are processed first
```

Comment:
Maybe it makes sense to have lower numbers be higher priority. That way we don't have priority inflation up to max int.